File: dailyQuestion/2020/2020-09/9-01/golang/solution_bfs.go
package main

import "fmt"

func main() {
	root := &TreeNode{Val: 3, Left: &TreeNode{Val: 5}, Right: &TreeNode{Val: 1}}
	root.Left.Left = &TreeNode{Val: 6}
	root.Left.Right = &TreeNode{Val: 2, Left: &TreeNode{Val: 7}, Right: &TreeNode{Val: 4}}
	root.Right.Left = &TreeNode{Val: 0}
	root.Right.Right = &TreeNode{Val: 8}

	result := pathSum(root, 12)
	fmt.Println(result)

}

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

type pair struct {
	node *TreeNode
	left int
}

func pathSum(root *TreeNode, sum int) (result [][]int) {
	if root == nil {
		return
	}
	parent := map[*TreeNode]*TreeNode{}
	var getPath func(node *TreeNode) (path []int)
	getPath = func(node *TreeNode) (path []int) {
		for ; node != nil; node = parent[node] {
			path = append(path, node.Val)
		}
		for i, j := 0, len(path)-1; i < j; i++ {
			path[i], path[j] = path[j], path[i]
			j--
		}
		return
	}
	stack := []pair{pair{node: root, left: sum}}
	for len(stack) != 0 {
		p := stack[0]
		stack = stack[1:]
		node := p.node
		left := p.left
		left -= node.Val
		if node.Left == nil && node.Right == nil {
			if left == 0 {
				result = append(result, getPath(node))
			}
		} else {
			if node.Left != nil {
				parent[node.Left] = node
				stack = append(stack, pair{node: node.Left, left: left})
			}
			if node.Right != nil {
				parent[node.Right] = node
				stack = append(stack, pair{node: node.Right, left: left})
			}
		}
	}
	return
}

File: dailyQuestion/2020/2020-09/9-01/python/solution_bfs.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

def pathSum(root: TreeNode, sum: int) -> [[int]]:
    result = []
    if not root :
        return result
    def getpath(node) :
        temp = []
        while node :
            temp.append(node.val)
            node = parent.get(node)
        result.append(temp[::-1])

    parent = {}
    stack = [{"node":root, "left":sum}]
    while stack :
        p = stack[0]
        stack = stack[1:]
        node = p["node"]
        left = p["left"]
        left -= node.val
        if not node.left and not node.right :
            if left == 0 :
                getpath(node)
        else :
            if node.left :
                parent[node.left] = node
                stack.append({"node":node.left,"left":left})
            if node.right :
                parent[node.right] = node
                stack.append({"node":node.right,"left":left})
    return result


if __name__ == "__main__" :
    node = TreeNode(3)
    node.left = TreeNode(5)
    node.right = TreeNode(1)
    node.left.left = TreeNode(6)
    node.left.right = TreeNode(2)
    node.left.right.left = TreeNode(7)
    node.left.right.right = TreeNode(4)

    node.right.left = TreeNode(0)
    node.right.right = TreeNode(8)
    result = pathSum(node,12)
    print(result)